### PR TITLE
Python, R, Julia and Javascript kernels are supported now !

### PR DESCRIPTION
**Pull request by Loghijiaha**
## Checklist

- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] Unit tests pass locally with my changes
- [ ] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes

## Types of changes

- [ ] Infrastructure change (non-breaking change which updates dependencies or improves infrastructure)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Further comments

javascript_build
javascript_output

### CODE REVIEW

File: src/main/resources/io/jenkins/plugins/ml/IPythonBuilder/config.jelly [hunk 1]
```diff
@@ -27,6 +27,15 @@
     <f:entry title="${%ML Task name}" field="task">
         <f:textbox name="task"/>
     </f:entry>
+    <f:entry title="${%Kernel}" field="kernel">
+        <select name="IPythonBuilder.kernel">
+            <option value="python" selected="${instance.kernel.equals('python')? 'true':null}">Python</option>
+            <option value="ir" selected="${instance.kernel.equals('ir')? 'true':null}">R</option>
+            <option value="julia-1.0" selected="${instance.kernel.equals('julia-1.0')? 'true':null}">Julia</option>
+            <option value="javascript" selected="${instance.kernel.equals('javascript')? 'true':null}">Javascript
+            </option>
```
[the remark this answers is not in the document]
Comment (Loghijiaha) [reply src/main/resources/io/jenkins/plugins/ml/IPythonBuilder/config.jelly:36]: We had to find a solution to select the kernel dynamically. But if the user wants to run pipeline jobs it is very easy to give the exact name of the kernel to run the script. Will discuss that.

File: src/main/java/io/jenkins/plugins/ml/IPythonKernelInterpreter.java [hunk 2]
```diff
@@ -99,18 +104,16 @@ public String toString() {
...
 
-    private static InterpreterContext getInterpreterContext() {
-        return new InterpreterContext.Builder()
-                .setNoteId("noteID")
+    private InterpreterContext getInterpreterContext(String kernel) {
+        Map<String, String> localProperties = new HashMap<>();
+        localProperties.put("kernel", kernel);
+        return InterpreterContext.builder()
+                .setNoteId("noteId")
                 .setParagraphId("paragraphId")
```
[the remark this answers is not in the document]
Comment (Loghijiaha) [reply src/main/java/io/jenkins/plugins/ml/IPythonKernelInterpreter.java:112]: Any string can be set to these parameters. Usually, These are attributes of a Zeppelin Notebooks. In Jupyter `note id` is `notebook name` and  `paragraph` is like `cell`.

**Comment by Loghijiaha** [comment]
@kinow Now all issues are resolved that you got before. Added test to checkMethod for code editor. Iteratively tested. Hope it will work.

**Comment by Loghijiaha** [comment]
global_config
Removed server address(not nessesary as we discussed) and add kernel. Other parameters should be defined.

File: src/main/resources/io/jenkins/plugins/ml/Server/help-kernel.html [hunk 3]
```diff
@@ -23,9 +23,9 @@
 -->
 
 <div>
-  IP address of the server
+  Appropriate kernel name
   <br>
-  No protocol will be needed for this.
+  Try `jupyter kernelspec list` to list out available kernels
```
Comment (kinow) [inline src/main/resources/io/jenkins/plugins/ml/Server/help-kernel.html:28]: :+1:  good idea! That might save you some issues/questions in issues/etc.

**Comment by kinow** [comment]
Ignore that @Loghijiaha , tried it again (as I noticed CI build was OK here), and it worked. I suspect that's because I had Eclipse starting at the same time, probably running another build.

**Comment by Loghijiaha** [comment]
I thought tht user will give the kernel name to select which language they want to run. The name will be usefull in tha job configuration to select which type of language they want. (there is a dropdown list to select configured kerenls in job config). This is same as how we configuring as a jupyter notebook.  I can migrate the option to build step instead of whole job configuration. Because admin should be aware of these kernel names. Users may not.

(Before the last commit, we had our kernel in buildstep. But  a hardcoded dropdown list.)

**Comment by kinow** [comment]
> I thought tht user will give the kernel name to select which language they want to run. The name will be usefull in tha job configuration to select which type of language they want. (there is a dropdown list to select configured kerenls in job config). This is same as how we configuring as a jupyter notebook. I can migrate the option to build step instead of whole job configuration. Because admin should be aware of these kernel names. Users may not.
> 
> (Before the last commit, we had our kernel in buildstep. But a hardcoded dropdown list.)

Yup, I think we can have the exact same screen as before. But instead of the hard-coded kernels, we can show the list of kernels from the global configurations.

Jenkins admins can still manage the kernel, and update if there are any changes. And users can either specify in their Jenkinsfile or job configuration which kernel each build step needs.

**Comment by kinow** [comment]
(ps: if a log entry is expected, and in a third party library, there are still ways to shush it and avoid cluttering the logs :+1: )

**Comment by kinow** [comment]
And CI error was on Windows, not important. Build passed fine on my PC.

**Comment by Loghijiaha** [comment]
I will keep the logs short in next PR. :+1:

**Pull request merged**